Trace.py

```diff
@@ -64,20 +64,19 @@ def _dump_pfb(xbox):
     return bytes(buffer)
 
 
-def _read_pgraph_rdi(xbox, offset, count):
+def _read_pgraph_rdi(xbox: Xbox, offset: int, count: int) -> bytes:
     # FIXME: Assert pusher access is disabled
     # FIXME: Assert PGRAPH idle
 
     NV10_PGRAPH_RDI_INDEX = 0xFD400750
     NV10_PGRAPH_RDI_DATA = 0xFD400754
 
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, offset)
-    data = bytearray()
-    for _ in range(count):
-        word = xbox.read_u32(NV10_PGRAPH_RDI_DATA)
-        data += struct.pack("<L", word)
+    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
+
+    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
+
+    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, original_offset)
```

Comment:
I'm not sure if it's even possible to restore the RDI index.
As far as I remember, the RDI has an internal state machine.

Reading the DATA register 4 times returns X,Y,Z,W (not in that order as far as I remember), but during that time the INDEX register will stay constant, only on the final read I think it auto-increments the INDEX.
If you touch the INDEX register, you may or may not be resetting the internal state machine.